File: Solutions/Remove_Covered_Intervals/Remove_Covered_Intervals.js
var removeCoveredIntervals = function (intervals) {
  intervals.sort((a, b) => {
    if (a[0] == b[0]) return b[1] - a[1];
    else return a[0] - b[0];
  });
  let remove = 0;
  for (let i = 0; i < intervals.length - 1; i++) {
    if (intervals[i][1] >= intervals[i + 1][1]) {
      remove++;
      intervals[i + 1] = intervals[i];
    }
  }
  return intervals.length - remove;
  // Time Complexity: O(nlog(n))
  // Space Complexity: O(1)
};

//Most common
/**
 * @param {number[][]} intervals
 * @return {number}
 */
var removeCoveredIntervals = function (intervals) {
  intervals.sort((a, b) => {
    if (a[0] == b[0]) return b[1] - a[1];
    else return a[0] - b[0];
  });
  let remove = 0;
  for (let i = 0; i < intervals.length - 1; i++) {
    if (intervals[i][1] >= intervals[i + 1][1]) {
      remove++;
      intervals[i + 1] = intervals[i];
    }
  }
  return intervals.length - remove;
};

//Second fastest
/**
 * @param {number[][]} intervals
 * @return {number}
 */
var removeCoveredIntervals = function (intervals) {
  let numCoverages = 0;
  for (let i = 0; i < intervals.length; i++) {
    for (let j = 0; j < intervals.length; j++) {
      if (i === j) continue;
      let a = intervals[i][0];
      let b = intervals[i][1];
      let c = intervals[j][0];
      let d = intervals[j][1];
      if (c <= a && b <= d) {
        numCoverages++;
        break;
      }
    }
  }
  return intervals.length - numCoverages;
};

//Fastest
/**
 * @param {number[][]} intervals
 * @return {number}
 */
var removeCoveredIntervals = function (intervals) {
  intervals.sort((a, b) => a[0] - b[0]);
  let count = intervals.length;
  let max = intervals[0][1];
  let maxStart = intervals[0][0];
  for (let i = 1; i < intervals.length; i++) {
    if (max >= intervals[i][1]) {
      count--;
    } else if (intervals[i][0] === maxStart) {
      count--;
      max = intervals[i][1];
    } else {
      max = intervals[i][1];
      maxStart = intervals[i][0];
    }
  }
  return count;
};

File: Solutions/Remove_Covered_Intervals/Remove_Covered_Intervals.go
func removeCoveredIntervals(intervals [][]int) int {
    sort.Slice(intervals, func(i, j int) bool {
        if intervals[i][0] != intervals[j][0] {
            return intervals[i][0] < intervals[j][0]
        } else {
            return intervals[i][1] > intervals[j][1]
        }
    })
    bound := -1
    valid := 0
    for _, v := range intervals {
        if v[1] <= bound {
            continue
        } else {
            bound = v[1]
            valid++
        }
    }
    return valid
}

//Another fastest
func removeCoveredIntervals(intervals [][]int) int {
	if len(intervals) <= 1 {
		return 0
	}
	sort.Slice(intervals, func(i, j int) bool {
		if intervals[i][0] == intervals[j][0] {
			return intervals[i][1] > intervals[j][1]
		} else {
			return intervals[i][0] < intervals[j][0]
		}
	})

	ans := 0
	minRight := intervals[0][1]
	for i := 1; i < len(intervals); i++ {
		if intervals[i][1] <= minRight {
			ans++
		} else {
			minRight = intervals[i][1]
		}
	}
	return len(intervals) - ans
}